### PR TITLE
Remove set_hs_protocol function

### PR DESCRIPTION
**Pull request by maskit**
This is a preparation for implementing Key Update and dropping keys. The goal of this PR is eliminating `set_hs_protocol`.

I'm going to introduce `QUICPacketProtectionKeyInfo` as a shared storage. It will be shared between `QUICHandshakeProtocol` and classes that use keys. By using the class, the key users don't need to access *a handshake protocol*.

This will simplify QUICHandshakeProtocol interface, and remove necessity of adding functions like `drop_key` to QUICHandshakeProtocol.

### CODE REVIEW

**Comment by maskit** [comment]
This is still WIP.

**Comment by maskit** [comment]
`KeyMaterial` and `QUICPacketProtection` were replaced with `QUICPacketProtectionKeyInfo`. Like I wrote on the description, this object is shared between a key generator and key users.

I also introduced `QUICPacketPayloadProtector`. Packet protection (payload encryption) was done by `QUICHandshakeProtocol` (`QUICTLS`), but now `QUICHandshakeProtocol` only does handshake. Keying materials generated as a result of handshake will be stored into `QUICPacketProtectionKeyInfo`, and `QUICPacketPaylaodProtector` (and `QUICPacketHeaderProtector`) will read the keys from it.

I'm still updating tests so this PR still has WIP label, but please start review. At minimum, a simple GET request from ngtcp was succeeded.

However there are many places we can optimize, I'm going to make small changes when I have time after I merge this. I think adding support for Key Update and dropping keys have priority (and anybody can work on these once we merge this).

**Comment by maskit** [comment]
It's ready for review.

**Comment by scw00** [comment]
Cool. Will have a look tomorrow.

File: iocore/net/quic/QUICPacketProtectionKeyInfo.cc [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+void
+QUICPacketProtectionKeyInfo::set_cipher_for_hp(const QUIC_EVP_CIPHER *cipher)
+{
+  this->_cipher_for_hp = cipher;
+}
+
+const uint8_t *
+QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase) const
```
Comment (scw00) [inline iocore/net/quic/QUICPacketProtectionKeyInfo.cc]:
```
uint8_t *
QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase)

const uint8_t *
QUICPacketProtectionKeyInfo::encryption_key_for_hp(QUICKeyPhase phase) const
```

Not sure do we need these two function.

Comment (maskit) [reply iocore/net/quic/QUICPacketProtectionKeyInfo.cc]: I wanted to make sure only classes that have the non-const object can update the information. By having the two, key *users* cannot break the information.

File: iocore/net/quic/QUICTLS.cc [hunk 2]
```diff
@@ -91,44 +89,69 @@ QUICTLS::is_ready_to_derive() const
...
+  uint8_t *server_iv;
+  size_t server_key_for_hp_len;
+  size_t server_key_len;
+  size_t *server_iv_len;
 
-  return 1;
-}
+  if (this->_netvc_context == NET_VCONNECTION_IN) {
+    client_key_for_hp     = this->_pp_key_info.decryption_key_for_hp(QUICKeyPhase::INITIAL);
+    client_key_for_hp_len = this->_pp_key_info.decryption_key_for_hp_len(QUICKeyPhase::INITIAL);
```
Comment (scw00) [inline iocore/net/quic/QUICTLS.cc]: don't understand when and where to set `_client_key_for_hp_len` in `pp_key_info` and what's value of client_key_for_hp_len  on this line.

Comment (scw00) [reply iocore/net/quic/QUICTLS.cc]: never mind . I found it

File: iocore/net/quic/QUICPacketProtectionKeyInfo.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  const QUIC_EVP_CIPHER *_cipher_initial = nullptr;
+  const QUIC_EVP_CIPHER *_cipher         = nullptr;
+  size_t _tag_len                        = 0;
+
+  bool _is_client_key_available[5] = {false};
+  bool _is_server_key_available[5] = {false};
+
+  // FIXME EVP_MAX_KEY_LENGTH and EVP_MAX_IV_LENGTH are not enough somehow
+  uint8_t _client_key[5][512];
```
Comment (masaori335) [inline iocore/net/quic/QUICPacketProtectionKeyInfo.h]: Just curious about why array of KeyMaterial doesn't work.

File: iocore/net/quic/QUICPacketProtectionKeyInfo.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+#pragma once
+
+#include "QUICTypes.h"
+#include "QUICKeyGenerator.h"
+
+class QUICPacketProtectionKeyInfo
+{
+public:
+  enum class Context { SERVER, CLIENT };
```
Comment (masaori335) [inline iocore/net/quic/QUICPacketProtectionKeyInfo.h]: Defining this to make `QUICPacketProtectionKeyInfo` independent from NetVC ?

**Comment by masaori335** [comment]
@maskit Do you have any Key Update code based on this?

File: iocore/net/quic/test/test_QUICPacketFactory.cc [hunk 5]
```diff
@@ -28,9 +28,9 @@
 
 TEST_CASE("QUICPacketFactory_Create_VersionNegotiationPacket", "[quic]")
 {
-  QUICPacketFactory factory;
-  MockQUICHandshakeProtocol hs_protocol;
-  factory.set_hs_protocol(&hs_protocol);
+  MockQUICPacketProtectionKeyInfo pp_key_info;
+  QUICPacketFactory factory(pp_key_info);
+  ;
```
Comment (masaori335) [inline iocore/net/quic/test/test_QUICPacketFactory.cc]: [nitpicks] `;`

File: iocore/net/quic/QUICPacketProtectionKeyInfo.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  // Payload Protection (common)
+
+  virtual const QUIC_EVP_CIPHER *get_cipher(QUICKeyPhase phase) const;
+  virtual size_t get_tag_len(QUICKeyPhase phase) const;
+  virtual void set_cipher_initial(const QUIC_EVP_CIPHER *cipher);
+  virtual void set_cipher(const QUIC_EVP_CIPHER *cipher, size_t tag_len);
+
+  // Payload Protection (encryption)
+
+  virtual bool is_encryption_key_available(QUICKeyPhase phase) const;
```
[the remark this answers is not in the document]
Comment (maskit) [reply iocore/net/quic/QUICPacketProtectionKeyInfo.h]: I needed to make them virtual for tests (making a mock class), and I didn't need to do the same on decryption side. But I'll add `virtual` consistently.

File: iocore/net/quic/QUICPacketPayloadProtector_openssl.cc [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+bool
+QUICPacketPayloadProtector::_unprotect(uint8_t *plain, size_t &plain_len, size_t max_plain_len, const uint8_t *cipher,
+                                       size_t cipher_len, uint64_t pkt_num, const uint8_t *ad, size_t ad_len, const uint8_t *key,
+                                       const uint8_t *iv, size_t iv_len, const EVP_CIPHER *aead, size_t tag_len) const
+{
+  uint8_t nonce[EVP_MAX_IV_LENGTH] = {0};
+  size_t nonce_len                 = 0;
+  _gen_nonce(nonce, nonce_len, pkt_num, iv, iv_len);
```
[the remark this answers is not in the document]
Comment (maskit) [reply iocore/net/quic/QUICPacketPayloadProtector_openssl.cc]: will do

File: iocore/net/quic/QUICPacketProtectionKeyInfo.h (hunk 3 above)
Comment (maskit) [reply iocore/net/quic/QUICPacketProtectionKeyInfo.h]: Not sure :) These are came from the old code, and IIRC, it seemed like buffer overflow.

File: iocore/net/quic/QUICPacketProtectionKeyInfo.h (hunk 4 above)
Comment (maskit) [reply iocore/net/quic/QUICPacketProtectionKeyInfo.h]: I didn't think like it, but I don't want the dependency.

**Comment by maskit** [comment]
> Do you have any Key Update or dropping keys code based on this?

No, I don't have any code at the moment.

I think next step for Key Update will be adding `QUICHandshake::get_key_genearator()` or passing a key generator to `QUICHandshakeProtocol` instead of `QUICPPKeyInfo`. In either ways, `QUICNetVC` will have a full and direct control for managing keys. It should satisfy key update and key drop scenarios. I may introduce `QUICKeyUpdater` to delete the processes.

**Pull request merged**